### PR TITLE
Add LoRA co-training support for HF EAGLE speculative decoding

### PR DESCRIPTION
**Pull request by yeyu-nvidia**
### What does this PR do?

Type of change: New feature + bug fixes

Adds **LoRA co-training** support for HF EAGLE speculative decoding. When `eagle_base_lora=True`, HF PEFT LoRA adapters are injected into the base model and co-trained alongside the EAGLE draft module in a single online training pass. A preservation loss (KL divergence between the original frozen base model output and the LoRA-adapted output) prevents base model drift. LoRA adapter weights are exported in standard peft format alongside EAGLE draft artifacts.

### Key features

- **LoRA injection**: `peft.inject_adapter_in_model` applied in-place (no wrapper), keeping the existing `HFEagleModel` structure intact.
- **Preservation loss**: Cross-entropy `H(ref, lora)` — equivalent gradient to `KL(ref || lora)` since `H(ref)` is constant w.r.t. LoRA params.
- **Warmup schedule**: `eagle_base_lora_warmup_steps` freezes LoRA for N steps while the EAGLE head stabilizes, then enables co-training via a `LoRAWarmupCallback`.
- **Logits detach regularization**: `eagle_base_lora_logits_detach_prob` stochastically detaches base logits from the EAGLE loss path, preventing LoRA from degenerating to maximize EAGLE accuracy at the cost of base model quality.
- **Export**: Standard peft format (`adapter_model.safetensors` + `adapter_config.json`) alongside EAGLE draft model.
- **Merge script**: `scripts/merge_lora.py` merges LoRA weights into the base model and restores the original `config.json` (avoids transformers 5.x rewriting `rope_theta` → `rope_parameters` which breaks vLLM/TRT-LLM).
- **Multinode fix**: `dp_shard_size` now uses `WORLD_SIZE` instead of local GPU count.

### Config options

```python
mtsp.convert(model, mode=[("eagle", {
    "eagle_base_lora": True,                          # enable LoRA co-training
    "eagle_base_lora_rank": 64,                       # LoRA rank
    "eagle_base_lora_alpha": 16.0,                    # LoRA scaling
    "eagle_base_lora_target_modules": ["q_proj", "k_proj", "v_proj", "o_proj"],
    "eagle_base_lora_preservation_loss_weight": 0.1,  # preservation loss weight
    "eagle_base_lora_warmup_steps": 0,                # freeze LoRA for N steps
    "eagle_base_lora_logits_detach_prob": 0.5,        # detach prob (0=never, 1=always)
})])
```

### Experimental results (Qwen3-8B, checkpoint-60000)

Base model quality preserved across detach_prob sweep (lm_eval: IFEval, ARC-C, Winogrande — results pending final collection).

**Acceptance rate** (mt_bench, draft_length=3, output_length=4096, temperature=0):

| detach_prob | vLLM AR | TRT-LLM AR |
|---|---|---|
| baseline (no LoRA) | 2.14 | 2.15 |
| 0.5 | 1.45 | 1.44 |
| 0.8 | **3.06** | **3.01** |
| 0.85 | 2.90 | 2.90 |
| 0.9 | 2.76 | 2.77 |
| 0.95 | 2.51 | 2.58 |
| 0.99 | 2.37 | 2.37 |
| 0.999 | 2.30 | 2.27 |
| 0.9999 | 2.31 | 2.26 |

Best AR at `detach_prob=0.8`: ~40% improvement over baseline.

### Testing

`tests/unit/torch/speculative/plugins/test_hf_speculative_lora.py` (5 tests):
- `test_lora_layers_injected` — LoRA layers present after conversion
- `test_trainable_params` — only `lora_*` and `eagle_module` params are trainable
- `test_forward_returns_loss` — forward returns non-zero scalar loss
- `test_eagle_offline_incompatible` — `eagle_base_lora=True` + `eagle_offline=True` raises `ValueError`
- `test_export_lora_artifacts` — export produces standard peft adapter files

### Bug fixes (included in this PR)

1. **`launch_train.sh` case pattern ordering**: glob `--eagle_base_lora*` was before specific patterns (`--eagle_base_lora_rank*`, etc.), silently swallowing LoRA args.
2. **LoRA optimizer exclusion during warmup**: warmup freezing excluded LoRA from the optimizer entirely; fixed with `add_param_group` in the callback.
3. **`merge_lora.py` config.json**: `save_pretrained()` with transformers >=5.x rewrites `rope_theta` → `rope_parameters`, breaking vLLM positional embeddings. Fixed by copying the original base model config.
4. **Multinode `dp_shard_size`**: used local GPU count instead of `WORLD_SIZE`.

### Checklist

- [x] Backward compatible (all new config fields have defaults)
- [x] Uses `peft` via lazy imports (no hard dependency)
- [x] Unit tests added
- [x] Online HF training only (`eagle_offline=True` blocked)

### CODE REVIEW

**Comment by h-guo18** [comment]
How would the base model quality and AL looks like with this lora cotraining?

**Comment by yeyu-nvidia** [comment]
> How would the base model quality and AL looks like with this lora cotraining?

Haven't tested. will report later

File: modelopt/torch/speculative/plugins/transformers.py [hunk 1]
```diff
@@ -37,6 +37,9 @@
 import torch
 import transformers
 from packaging.version import Version
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/plugins/transformers.py]: **Blocker:** These top-level `from peft import ...` lines make `peft` a hard dependency for the entire transformers plugin. Anyone importing this module without `peft` installed will get an `ImportError`, even if they don't use LoRA.

These must be lazy imports inside `_inject_base_lora`, `_set_base_lora_enabled`, and the export method. For example:
```python
def _inject_base_lora(self):
    from peft import LoraConfig
    from peft.mapping import inject_adapter_in_model
    ...

def _set_base_lora_enabled(self, enabled: bool):
    from peft.tuners.lora import LoraLayer
    ...
```

File: modelopt/torch/speculative/config.py [hunk 2]
```diff
@@ -110,3 +110,37 @@ class EagleConfig(ModeloptBaseConfig):
...
+
+    eagle_base_lora_rank: int = ModeloptField(
+        default=64,
+        description="LoRA rank for the base model adapters.",
+    )
+
+    eagle_base_lora_alpha: float = ModeloptField(
+        default=16.0,
+        description="LoRA alpha (scaling) for the base model adapters.",
+    )
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/config.py:214]: Mutable default: `default=[]` means all config instances share the same list object. Use `default_factory=list` or `default=None` with a note that `None` uses peft defaults.

File: modelopt/torch/speculative/plugins/transformers.py [hunk 3]
```diff
@@ -547,6 +550,37 @@ def _get_eagle_device(self):
...
+        )
+        inject_adapter_in_model(lora_config, self._base_model, adapter_name="default")
+        # Unfreeze only the LoRA parameters
+        for name, param in self._base_model.named_parameters():
+            if "lora_" in name:
+                param.requires_grad = True
+
+    def _set_base_lora_enabled(self, enabled: bool) -> None:
+        """Enable or disable LoRA adapters in the base model."""
+        for module in self._base_model.modules():
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/plugins/transformers.py:600]: The docstring says "KL divergence" but this computes cross-entropy: `-softmax(ref) * log_softmax(lora)`. The missing entropy term is constant w.r.t. LoRA params so gradients are correct, but the naming is misleading. Either rename or add a comment clarifying this is KL up to a constant.

File: modelopt/torch/speculative/plugins/transformers.py [hunk 4]
```diff
@@ -723,7 +762,9 @@ def _compute_ttt_attention_mask(
     ) -> BlockMask | torch.Tensor:
         """Return TTT attention_mask tensor of type BlockMask or Tensor depends on eagle attn impl."""
         msk_func = get_ttt_msk_func(seq_length, ttt_step)
-        dtypemin = torch.finfo(self._base_llm_config.dtype).min
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/plugins/transformers.py]: The reference forward with LoRA disabled has no `try/finally`. If the forward throws, LoRA stays disabled for all subsequent calls:
```python
self._set_base_lora_enabled(False)
try:
    ref_logits = _run_forward(no_grad=True).logits
    if hasattr(self, "_aux_hidden_states"):
        self._aux_hidden_states.clear()
finally:
    self._set_base_lora_enabled(True)
```

File: modelopt/torch/speculative/plugins/transformers.py [hunk 5]
```diff
@@ -610,6 +644,11 @@ def modify(
                 if layer_idx in self.eagle_config.eagle_aux_hidden_state_layer_ids:
                     layer.register_forward_hook(self._collect_aux_hidden_states_forward_hook)
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/plugins/transformers.py:681]: `assert not self.eagle_offline` can be optimized out with `python -O`. Use `if self.eagle_offline: raise ValueError(...)` for runtime config validation.

File: modelopt/torch/export/plugins/hf_spec_export.py [hunk 6]
```diff
@@ -112,11 +112,17 @@ def _check_valid_sd(self, export_sd: dict):
             "llama": LLAMA_EAGLE_SINGLE_LAYER,
             "kimik2": KIMIK2_EAGLE_SINGLE_LAYER,
         }[self.eagle_decoder_type]
+        # fc and hidden_norm are only present when use_aux_hidden_state=True
+        use_aux = getattr(self.model.eagle_config, "use_aux_hidden_state", False)
+        aux_only_keys = {"fc", "layers.0.hidden_norm"}
+        required_keys = set(expected_keys_single_layer["required"])
+        if not use_aux:
+            required_keys -= aux_only_keys
         # Check that export sd has required keys
```
Comment (ChenhanYu) [inline modelopt/torch/export/plugins/hf_spec_export.py:125]: LoRA key filtering (`"lora_A" in k or "lora_B" in k`) is fragile. PEFT may use other key patterns (`lora_embedding_A`, `lora_magnitude_vector`, etc.). Consider using peft's own utilities to identify adapter parameters, or add a warning if zero LoRA tensors are found.

File: modelopt/torch/speculative/utils.py [hunk 7]
```diff
@@ -467,7 +467,7 @@ def enable_cp_ttt_patch():
     import modelopt.torch.speculative.plugins.transformers
 
     modelopt.torch.speculative.plugins.transformers.ENABLE_CP_TTT_PATCH = True
-    with sdpa_kernel(SDPBackend.CUDNN_ATTENTION):
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/utils.py:560]: Adding `SDPBackend.MATH` fallback and the `dtype` getattr changes in `transformers.py` look unrelated to LoRA co-training. Consider splitting into a separate PR or calling them out in the description.

**Review by ChenhanYu — CHANGES_REQUESTED** [review]
This PR adds LoRA co-training support for HF EAGLE speculative decoding — reasonable feature design with good test coverage. However, **`peft` is now a hard import-time dependency** for the entire `transformers.py` plugin due to top-level imports. This will break anyone importing `modelopt.torch.speculative.plugins.transformers` without `peft` installed, even if they don't use LoRA. This needs to be fixed before merging. Several other issues below.

File: modelopt/torch/speculative/plugins/transformers.py (hunk 1 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/plugins/transformers.py]: peft is not imported at the top-level

File: modelopt/torch/speculative/plugins/transformers.py (hunk 3 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/plugins/transformers.py:600]: This is logit KL divergence

File: modelopt/torch/speculative/utils.py (hunk 7 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/utils.py:560]: This is needed for tests

File: modelopt/torch/speculative/plugins/transformers.py (hunk 1 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/plugins/transformers.py]: Fixed in 5632abab — peft imports moved inside `_inject_base_lora()` (`LoraConfig`, `inject_adapter_in_model`) and `_set_base_lora_enabled()` (`LoraLayer`), so peft is no longer a hard top-level dependency.

File: modelopt/torch/speculative/config.py (hunk 2 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/config.py:214]: Fixed in 5632abab — changed to `list | None = ModeloptField(default=None)`. All existing usages already used `or None` so nothing breaks.

File: modelopt/torch/export/plugins/hf_spec_export.py (hunk 6 above)
Comment (yeyu-nvidia) [reply modelopt/torch/export/plugins/hf_spec_export.py:125]: Fixed in 5632abab — tightened the filter to `.lora_A.` / `.lora_B.` (dot-bounded) and added a `RuntimeError` if no LoRA tensors are found.

File: modelopt/torch/export/plugins/hf_spec_export.py [hunk 8]
```diff
@@ -185,6 +186,20 @@ def _get_config_from_draft_or_base(key: str, model: nn.Module):
 
         return template_config
 
+    def _export_lora(self, export_dir: Path, full_sd: dict):
+        """Export base model LoRA adapter weights alongside the eagle module artifacts."""
+        lora_sd = {k: v for k, v in full_sd.items() if "lora_A" in k or "lora_B" in k}
+        save_file(lora_sd, export_dir / "lora_adapter_model.safetensors")
+
```
[the remark this answers is not in the document]
Comment (yeyu-nvidia) [reply modelopt/torch/export/plugins/hf_spec_export.py]: Fixed in 5632abab — added fail-fast `RuntimeError` when no LoRA tensors are found.

File: modelopt/torch/speculative/plugins/transformers.py (hunk 3 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/plugins/transformers.py:600]: To expand on this: the implementation computes cross-entropy `H(ref, lora) = -softmax(ref) · log_softmax(lora)`, which equals KL(ref ∥ lora) + H(ref). Since H(ref) is constant w.r.t. LoRA parameters, the gradients are identical to true KL divergence — so the optimization objective is equivalent. The docstring has been updated to clarify this: it now reads "KL(softmax(ref) || log_softmax(lora))" and notes that the entropy term of the reference is constant and dropped.

File: modelopt/torch/speculative/utils.py (hunk 7 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/utils.py:560]: To expand: the `SDPBackend.MATH` fallback in `enable_cp_ttt_patch()` is required for CPU unit tests. `CUDNN_ATTENTION` is only available on GPU, so without `MATH` as a fallback the test environment raises an error when no supported SDPA backend is found. This is directly exercised by the LoRA co-training forward pass test (`test_forward_returns_loss`), which runs the TTT attention path on CPU. The change is scoped to the `enable_cp_ttt_patch()` context manager and doesn't affect production GPU paths.

File: modelopt/torch/speculative/plugins/transformers.py (hunk 5 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/plugins/transformers.py:681]: Fixed in 6e709b55 — replaced with `if self.eagle_offline: raise ValueError(...)`. Also updated the test to expect `ValueError` instead of `AssertionError`.

File: modelopt/torch/speculative/plugins/transformers.py (hunk 4 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/plugins/transformers.py]: Fixed in 6e709b55 — wrapped the reference forward in `try/finally` so `_set_base_lora_enabled(True)` is guaranteed to run even if the forward throws.

**Review by ChenhanYu — APPROVED** [review]
Previous review feedback addressed: peft imports are now lazy (inside functions), mutable default fixed (None), try/finally for LoRA disable/re-enable added, assert replaced with raise ValueError. Two minor items remain (non-blocking): _preservation_loss docstring says "KL" but computes cross-entropy (gradients correct, naming misleading), and LoRA key filtering still uses lora_A/lora_B only (RuntimeError on zero keys mitigates). LGTM.

**Comment by h-guo18** [comment]
Idea: Separate this feature as a different optimization with vanilla Eagle. One approach in my mind:

Refactor LoraCotrainSpecModel as a subclass that inherits from HFEagleModel. Most Eagle logics reused, with a few methods overwritten/extended:
- modify
- _base_model_forward
- _collect_aux_hidden_states_forward_hook
- _inject_base_lora, _set_base_lora_enabled, _preservation_loss

LoraCotrainConfig should also inherit from base SpecConfig (like EagleConfig). 

Create a separate registry for LoraCotrainSpecModel. During conversion, convert to `HFEagleModel` or `LoraCotrainSpecModel` depends on user input. Similar to distillation [implemenmtation](https://github.com/NVIDIA/Model-Optimizer/blob/main/modelopt/torch/distill/mode.py) 

Bnefits:
- Core Eagle loigic remains minimal and clean, easier to maintain and hack on
- We can reuse this lora feature to Dflash/medusa with little efforts.

**Comment by yeyu-nvidia** [comment]
Thanks for the suggestion! I considered this but decided against the subclass approach for a few reasons:

1. **LoRA co-training is a training-time option, not a separate optimization mode.** It doesn't change the model architecture or inference behavior — the exported artifacts are identical (EAGLE draft head + standard peft adapter). Making it a separate registry entry implies it's a different spec decoding algorithm, when it's really just a training knob.

2. **The code coupling is minimal.** The LoRA logic is concentrated in `_inject_base_lora`, `_set_base_lora_enabled`, and `_preservation_loss` — all gated behind `eagle_base_lora=True`. The core Eagle forward/loss path is untouched when LoRA is off. A subclass would need to override `modify()` and `forward()` just to inject a few extra lines, duplicating 90%+ of the parent methods.

3. **Subclass fragility.** If `HFEagleModel.forward()` or `modify()` changes, the subclass overrides silently diverge. With the current approach, LoRA is woven into the same code path — any changes to Eagle automatically include LoRA handling. This is safer for maintenance.

4. **Extending to DFlash/Medusa doesn't require a subclass hierarchy.** The LoRA injection (`_inject_base_lora`) operates on the base model, not the draft head — it's agnostic to the spec decoding method. If we want to add LoRA to DFlash, we can extract the LoRA methods into a mixin or utility rather than building a parallel class tree.
